File: Data Structures & Algorithms/products-of-array-discluding-self/submission-4.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ans = new int[nums.length];
        int zeros = 0;
        for(int x : nums) {
            if(x==0) zeros++;
        }
        if(zeros>1) return ans;
        ans[0] = 1;
        for(int i=1; i<nums.length; i++) {
            ans[i] = ans[i-1] * nums[i-1];
        }
        int tmp = 1;
        for(int i=nums.length-1; i>=0; i--) {
            ans[i] *= tmp;
            tmp *= nums[i];
        }
        return ans;
    }
}  

File: Data Structures & Algorithms/permutations/submission-4.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        def helper(vis: set(), curr: List[int]):
            if len(curr)==len(nums):
                ans.append(curr.copy())
                return
            for i in range(len(nums)):
                if nums[i] in vis:
                    continue
                curr.append(nums[i])
                vis.add(nums[i])
                helper(vis, curr)
                curr.pop()
                vis.remove(nums[i])

        helper(set(), [])
        return ans